adr/merk-proofs.md

```diff
@@ -0,0 +1,725 @@
+```
+
+********Explanation********
+
+- We only care about node 1
+- Starting at node 2, we realise we don’t need it’s key value, but something we care about (node 1) might be to the left of it. Return the Node::KvHash (case 2)
+- At node 1, we care about its key and value so we return Node::KV (case 1)
+- At node 4, we don’t care about its value and none of the nodes we care about are its descendants so return Node::Hash (case 3)
+
+Let’s calculate the root hash of the truncated tree.
+
+We start by calculating the kv_hash and node_hash for node 1. 
+
+Since it’s a leaf node the node_hash is equal to its kv_hash.
+
+```mermaid
+graph TD;
+2["kv_hash: 2b"];
+2-->1["key: 1, value: a \n kv_hash = 1a \n node_hash = 1a \n ."];
+2-->4["node_hash: 4d3c5e"];
+```
+
+Finally we calculate the node_hash for the root node (doubles as the root_hash of the structure)
+
+```
+kv_hash = 2b
+left_node_hash = 1a
+right_node_hash = 4d3c5e
+
+node_hash = Hash(kv_hash, left_node_hash, right_node_hash)
+node_hash = Hash(2b, 1a, 4d3c5e)
+node_hash = 2b1a4d3c5e
+```
+
+Exactly the same as before!
+
+In the next section, we specify a powerful system for describing the keys we want from a state tree.
+
+****************************************Query System****************************************
+
+A query is a set of query items that are used to describe keys in a tree. In merk we have 10 query item types. 
+
+- Key(key_name)
+    - selects a node with the given key name
+- Range(start..end)
+    - selects all nodes that fall in this range with the end non-inclusive
+- RangeInclusive(start..=end)
+    - selects all nodes whose key fall in this range with the end inclusive
+- RangeFull(..)
+    - selects all nodes in the tree
+- RangeFrom(start..)
+    - selects all nodes with keys greater than or equal to start
+- RangeTo(..end)
+    - selects all nodes with keys less than the end (the end is not inclusive)
+- RangeToInclusive(..=end)
+    - selects all nodes with keys less than or equal to the end (end inclusive)
+- RangeAfter(prev..)
+    - select all nodes with keys greater than the prev
+- RangeAfterTo(prev..end)
+    - selects all nodes with keys greater than prev but less than the end
+- RangeAfterToInclusive(prev..=end)
+    - selects all nodes with keys greater than prev but less than or equal to the end
+
+We can then combine the query items above to build a query that selects non-contiguous keys e.g 
+
+- query = [QueryItem::Key(2), QueryItem::RangeInclusive(4..=5)]
+    - this selects 3 nodes with keys [2, 4, 5]
+
+**Query Constraints**
+
+- A query must have all its query items ordered by key
+- Overlapping query items must be merged
+
+**Example**
+
+query = [QueryItem::Range(2..4), QueryItem::RangeInclusive(3..=5)]
+
+QueryItem::Range(2..4) selects for [2, 3]
+
+QueryItem::RangeInclusive(3..=5) selects for: [3. 4. 5]
+
+an overlap exists, we merge the sets: [2, 3, 4, 5]
+
+This is equivalent to QueryItem::RangeInclusive(2..=5), hence the query becomes
```

Comment:
> becomes

It's unclear to me if a query optimizer fixes an overlapping query for us or it will return an error in case of overlap and we should fix it ourselves?

Reply:
Yes it does, we check if there is an overlap then we merge. 
See: https://github.com/dashpay/grovedb/blob/master/merk/src/proofs/query/insert.rs#L151